File: src/make-two-consecutive.js
import {
  groupWith,
  equals,
  compose,
  any,
  length,
  range,
  map,
  remove,
  join,
  __
} from 'ramda'

export const toPossibility = boolean =>
  boolean
    ? 'Possible'
    : 'Impossible'

export const isBeautiful = compose(
  any(group => length(group) >= 2),
  groupWith(equals)
)

export const allMostPerms = string =>
  map(
    compose(
      join(''),
      remove(__, 1, string)
    ),
    compose(
      range(0),
      length
    )(string)
  )

export const canMakeBeautiful = compose(
  toPossibility,
  any(isBeautiful),
  allMostPerms
)

File: src/haskell/make-two-consecutive.hs
import Data.List

data Possibility = Possible | Impossible
  deriving (Show)

toPossibility :: Bool -> Possibility
toPossibility True = Possible
toPossibility False = Impossible

isBeautiful :: Eq a => [a] -> Bool
isBeautiful = any ((>= 2) . length) . group

missingOnes :: [a] -> [[a]]
missingOnes list = [(take i list) ++ (drop (i + 1) list) | i <- [0..((length list) - 1)]]

couldBeBeautiful :: Eq a => [a] -> Possibility
couldBeBeautiful list = toPossibility (any isBeautiful (missingOnes list))
